test/pubsub/operation.node.js

```diff
@@ -0,0 +1,174 @@
+'use strict'
+/* eslint-env mocha */
+
+const chai = require('chai')
+chai.use(require('dirty-chai'))
+const { expect } = chai
+const sinon = require('sinon')
+
+const pWaitFor = require('p-wait-for')
+const pDefer = require('p-defer')
+const mergeOptions = require('merge-options')
+
+const multiaddr = require('multiaddr')
+const PeerInfo = require('peer-info')
+
+const { create } = require('../../src')
+const { subsystemOptions, subsystemMulticodecs } = require('./utils')
+
+const listenAddr = multiaddr('/ip4/127.0.0.1/tcp/0')
+const remoteListenAddr = multiaddr('/ip4/127.0.0.1/tcp/0')
+
+describe('Pubsub subsystem operates correctly', () => {
+  let peerInfo, remotePeerInfo
+  let libp2p, remoteLibp2p
+  let remAddr
+
+  beforeEach(async () => {
+    [peerInfo, remotePeerInfo] = await Promise.all([
+      PeerInfo.create(),
+      PeerInfo.create()
+    ])
+
+    peerInfo.multiaddrs.add(listenAddr)
+    remotePeerInfo.multiaddrs.add(remoteListenAddr)
+  })
+
+  describe('pubsub started before connect', () => {
+    beforeEach(async () => {
+      libp2p = await create(mergeOptions(subsystemOptions, {
+        peerInfo
+      }))
+
+      remoteLibp2p = await create(mergeOptions(subsystemOptions, {
+        peerInfo: remotePeerInfo
+      }))
+
+      await libp2p.start()
+      await remoteLibp2p.start()
+
+      remAddr = remoteLibp2p.transportManager.getAddrs()[0]
+    })
+
+    afterEach(() => Promise.all([
+      libp2p && libp2p.stop(),
+      remoteLibp2p && remoteLibp2p.stop()
+    ]))
+
+    afterEach(() => {
+      sinon.restore()
+    })
+
+    it('should get notified of connected peers on dial', async () => {
+      sinon.spy(libp2p.registrar, 'onConnect')
+      sinon.spy(remoteLibp2p.registrar, 'onConnect')
+
+      const connection = await libp2p.dialProtocol(remAddr, subsystemMulticodecs)
+
+      expect(connection).to.exist()
+      expect(libp2p.pubsub._pubsub.peers.size).to.be.eql(1)
+      expect(remoteLibp2p.pubsub._pubsub.peers.size).to.be.eql(1)
+
+      expect(libp2p.registrar.onConnect.callCount).to.equal(1)
+      expect(remoteLibp2p.registrar.onConnect.callCount).to.equal(1)
+    })
+
+    it('should receive pubsub messages', async () => {
```

Comment:
Is this test redundant with the implementations test? Looks to be validating the same things.